keps/sig-instrumentation/2831-kubelet-tracing/README.md

```diff
@@ -0,0 +1,354 @@
+
+### Running the OpenTelemetry Collector
+
+The [OpenTelemetry Collector](https://github.com/open-telemetry/opentelemetry-collector) can be run as a sidecar, a daemonset, a deployment , or a combination in which the daemonset buffers telemetry and forwards to the deployment for aggregation (e.g. tail-base sampling) and routing to a telemetry backend.  To support these various setups, the kubelet should be able to send traffic either to a local (on the control plane network) collector, or to a cluster service (on the cluster network).
+
+### Kubelet Configuration
+
+Refer to [OpenTelemetry Specification](https://github.com/open-telemetry/opentelemetry-specification/blob/ccbfdd9020ba0d581aa5d880235302f8b0eb8669/specification/resource/semantic_conventions/README.md#service)
+
+```golang
+// KubeletConfiguration contains the configuration for the Kubelet
+type KubeletConfiguration struct {
+    metav1.TypeMeta `json:",inline"`
+    ------------
+
+    // +optional
+    // TracingConfiguration
+    TracingConfiguration componentbaseconfig.TracingConfiguration
+}
+
+// TracingConfiguration specifies configuration for opentelemetry tracing
+type TracingConfiguration struct {
+    // +optional
+    // URL of the collector that's running on the node.
+    // Defaults to 0.0.0.0:4317
+    URL  *string `json:"url,omitempty" protobuf:"bytes,1,opt,name=url"`
+
+    // +optional
+    // SamplingRatePerMillion is the number of samples to collect per million spans.
+    // Defaults to 0.
+    SamplingRatePerMillion int32 `json:"samplingRatePerMillion,omitempty" protobuf:"varint,2,opt,name=samplingRatePerMillion"`
+}
+
+```
+If a TracingConfiguration is not specified, the kubelet will not send any spans, even if incoming requests ask for sampling.
+
+## Design Details
+
+### Test Plan
+
+We will test tracing added by this feature with an integration test.  The
+integration test will verify that spans exported by the kubelet match what is
+expected from the request. We will also add an integration test that verifies
+spans propagated from kubelet to API server match what is expected from the request.
+
+### Graduation Requirements
+
+Alpha
+
+- [] Implement tracing of incoming and outgoing gRPC, HTTP requests in the kubelet
+- [] Integration testing of tracing
+
+Beta
+
+- [] Publish examples of how to use the OT Collector with kubernetes
+- [] Allow time for feedback
+- [] Revisit the format used to export spans.
+
+GA
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+  - [X] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: KubeletTracing
+    - Components depending on the feature gate: kubelet
+  - [X] Other
+    - Describe the mechanism: Kubelet Configuration.
```

Comment:
If no TracingConfiguration is specified, the feature is effectively disabled because the SamplingRatePerMillion defaults to 0, right? Can we call that out explicitly somewhere?

Reply:
It is actually slightly different to disable tracing entirely vs setting sampling rate to 0.  The difference is that with a sampling rate of 0 and enabled, a trace context with sampled=true will still cause traces to be produced.  The other difference is whether or not the kubelet attempts to connect via otlp protocol to export traces.

Reply:
Okay, let's ensure this is clear in the PRR questionnaire. Thanks for the clarification!

Reply:
clarified that, thank you